### PR TITLE
core: split out manager_get_units_for_pidref() and use it at multiple places

### PR DESCRIPTION
**Pull request by poettering**
Split out of #32441

### CODE REVIEW

File: src/core/manager.c [hunk 1]
```diff
@@ -2827,41 +2858,26 @@ static int manager_dispatch_sigchld(sd_event_source *source, void *userdata) {
...
-                }
-                if (u2)
-                        manager_invoke_sigchld_event(m, u2, &si);
-                if (array_copy)
-                        for (size_t i = 0; array_copy[i]; i++)
-                                manager_invoke_sigchld_event(m, array_copy[i], &si);
+                        /* Finally, execute them all. Note that the array might contain duplicates, but that's fine,
+                         * manager_invoke_sigchld_event() will ensure we only invoke the handlers once for each
+                         * iteration. */
+                        for (int i = 0; i < n_array; i++)
```
Comment (YHNdnzj) [inline src/core/manager.c]: FOREACH_ARRAY?

**Review by yuwata — APPROVED** [review]

File: src/core/manager.c [hunk 2]
```diff
@@ -2613,6 +2613,59 @@ static void manager_invoke_notify_message(
...
+                         * the process currently is. Note that we do not bother with filtering duplicates
+                         * here. */
+
+                        size_t i = 0;
+                        if (u1)
+                                units[i++] = u1;
+                        if (u2)
+                                units[i++] = u2;
+                        if (array)
+                                for (size_t j = 0; array[j]; j++)
```
Comment (yuwata) [inline src/core/manager.c:2656]: FOREACH_ARRAY ?

Comment (poettering) [reply src/core/manager.c:2656]: hmm, we don't really know the length here easily. the end condition is a NULL check after all, not a size check...

Comment (yuwata) [reply src/core/manager.c:2656]: Ah, indeed.

File: src/core/manager.c [hunk 3]
```diff
@@ -2613,6 +2613,59 @@ static void manager_invoke_notify_message(
         }
 }
 
+static int manager_get_units_for_pidref(Manager *m, const PidRef *pidref, Unit ***ret_units) {
+        /* Determine array of every unit that is interested in the specified process */
+
+        assert(m);
+        assert(pidref);
```
Comment (YHNdnzj) [inline src/core/manager.c]: Hmm, maybe `assert(pidref_is_set(...))`?

File: src/core/manager.c [hunk 4]
```diff
@@ -2827,41 +2857,28 @@ static int manager_dispatch_sigchld(sd_event_source *source, void *userdata) {
...
+                         * iteration. */
+                        FOREACH_ARRAY(u, array, n_array)
+                                manager_invoke_sigchld_event(m, *u, &si);
                 }
-                if (u2)
-                        manager_invoke_sigchld_event(m, u2, &si);
-                if (array_copy)
-                        for (size_t i = 0; array_copy[i]; i++)
-                                manager_invoke_sigchld_event(m, array_copy[i], &si);
+
```
Comment (YHNdnzj) [inline src/core/manager.c]: spurious newline?

**Review by YHNdnzj — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Pull request merged**